### ISSUE TITLE
Visualize features in CNN+LSTM

### ISSUE THREAD

**Issue by gabarlacchi** [OverLordGoldDragon/see-rnn#4]
Hello!
Great work, i really appreciate it!
I am developing a network for video sequence classification composed by: 
CNN (for feature extraction) + LSTM (for sequence class prediction).
The features of each sequence (let's consider 30 frames) are given from a pre-trained network  fed into the LSTM with return_sequence=False.
Is possible to visualize the features that let the classification be possible, in a heatmap form?
I am facing with some troubles understanding the way of do that with your tool, even if it's seems very powerful.
Thank you for the help, 
Best

**Comment by OverLordGoldDragon** [comment]
Thanks

What you're looking for is a form of _attention_ that highlights parts of input that the network deems "important" for classification; that isn't something this package can do. Better repos for this are [iNNvestigate](https://github.com/albermax/innvestigate), and [Keras-vis](https://raghakot.github.io/keras-vis/visualizations/activation_maximization/) w/ saliency maps. They offer tools to perform the necessary transformations - once obtained, you can feed them to the [`show_features_`](https://github.com/OverLordGoldDragon/see-rnn/blob/master/see_rnn/visuals_gen.py) functions, or use their own tools.

**Comment by gabarlacchi** [comment]
Thanks for the reply.
I was too trivial on the first post.
I have tried the links that you showed up to me, but none of them are fit my goal.
What I am trying to do is to apply a sort of grad cam method to every input of my sequence. I post the code to better explanation, I really appreciate any comment by you though.

`
video = tf.keras.layers.Input(shape=(5, 300, 300, 3))
cnn_base = tf.keras.applications.Xception(input_shape=(300, 300, 3), weights='imagenet', include_top=False)
cnn_out = tf.keras.layers.GlobalAveragePooling2D()(cnn_base.output)
cnn = tf.keras.Model(inputs=cnn_base.input, outputs=cnn_out)
encoded_frames = tf.keras.layers.TimeDistributed(cnn)(video)
encoded_sequence = tf.keras.layers.LSTM(256)(encoded_frames)
outputs = tf.keras.layers.Dense(units=classes, activation='softmax')(hidden_layer)
model = tf.keras.Model(inputs=[video], outputs=outputs)
`
This could be a simple sequence-classification model. The images are fed into the cnn (Xception now, but whatever) and the features are extracted. Then they're used to make classification in some way.
I would like to catch the gradients of the layers of the cnn, meanwhile they are doing feature extraction. My idea it's to  save the gradients of the cnn's (applied to each image of a sequence by time distributed wrapper), and later on use those gradients to compute grad cam by every single image. 

`
x = np.zeros((1, 5,300,300,3)) 
y = tf.keras.utils.to_categorical(2, nu_classes=classes)
with tf.GradientTape() as tape:
        prediction = model(x, training=True)
        loss = tf.keras.losses.CategoricalCrossentropy(prediction, y)
        gradients = tape.gradient(loss, model.trainable_variables)
`
Something like the code above (very general to model), but applied to the gradients that are feeding into the LSTM, since they are suppose to be the gradients of the cnn output at i-th time step. And then saving them for each one. Once they are saved I can apply grad cam using some loss.
With grad cam I mean any method which can show me up an heatmap with activations (as in normal single-input cnn), but what is important is the calculation of the gradients within each step.

I know that is a kind of attention map, but it's exactly ant I want to verify. 
Is that can works? 
Your package can help in some way?
Thanks for any help

**Comment by OverLordGoldDragon** [comment]
If you seek code for simple gradient extraction, you can have a look at [`get_layer_gradients()`](https://github.com/OverLordGoldDragon/see-rnn/blob/master/see_rnn/inspect_gen.py#L47) - but most of the work comes afterwards, in manipulating the gradients into something meaningful. Sometimes an entire train routine is designed around gradient exploration, as in Keras' [Convolutional filters](https://keras.io/examples/conv_filter_visualization/) example.

This repository, however, doesn't offer any such introspection methods - though it's on a to-do list, just not anytime soon. If what I linked didn't make the fit, I can suggest a few more:

 - [Attention Mechanisms](https://github.com/uzaymacar/attention-mechanisms)
 - [Keras Attention](https://github.com/datalogue/keras-attention)
 - [LRP Toolbox](https://github.com/sebastian-lapuschkin/lrp_toolbox)
 - [LRP for LSTM](https://github.com/ArrasL/LRP_for_LSTM)

Lastly, I refer to the recent textbook around which LRP is based: [Explainable AI: Interpreting, Explaining, and Visualizing Deep Learning](https://sci-hub.se/10.1007/978-3-030-28954-6).

**Comment by jnthnroy** [comment]
Thank for the information, I have a lot of new toy now to explore...
keep going!

Jonathan

**Comment by OverLordGoldDragon** [comment]
Interpretability visualizations are on the possible to-do list. Assuming resolved.

**Comment by sepehrgolriz** [comment]
Dear @barloccia 

Could you find solution for your problem? I have the same problem as yours. Your kind help would be appreciated if you could help me in this regard.
